pkg/tracing/tracing.go

```diff
@@ -0,0 +1,157 @@
+	traceNoop "go.opentelemetry.io/otel/trace/noop"
+)
+
+type customExporter struct {
+	exporter     *otlptrace.Exporter
+	failureCount int
+	lock         sync.Mutex
+}
+
+// Shutdown implements trace.SpanExporter.
+func (ce *customExporter) Shutdown(ctx context.Context) error {
+	log.Info("Shutting down the exporter")
+	return ce.exporter.Shutdown(ctx)
+}
+
+func (ce *customExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
+	err := ce.exporter.ExportSpans(ctx, spans)
+	ce.lock.Lock()
+	defer ce.lock.Unlock()
+
+	if err != nil {
+		ce.failureCount++
+		log.Errorf("Traces export failed: %v", err)
+
+		// Check if the failure threshold is exceeded.
+		if ce.failureCount >= 3 {
+			fmt.Println("Disabling tracing due to repeated export failures.")
+			log.Errorf("Disabling tracing due to repeated export failures.")
+			config.SetTracingEnabled(false)
+
+			// Set the tracer provider to a no-op provider. This will prevent any further spans from being exported.
+			otel.SetTracerProvider(traceNoop.NewTracerProvider())
+
+			// Shut down the custom exporter to clean up resources.
+			if shutdownErr := ce.Shutdown(ctx); shutdownErr != nil {
+				log.Errorf("Failed to shut down the tracer provider: %v", shutdownErr)
+			}
+		}
+	} else {
+		// Reset failure count on a successful export.
+		ce.failureCount = 0
+	}
+
+	return err
+}
+
+// InitTracing initializes the OpenTelemetry tracing.
+func InitTracing(serviceName string) func() {
+
+	if !config.IsTracingEnabled() {
+		log.Info("Tracing is disabled")
+		return func() {}
+	}
+
+	if serviceName == "" {
+		log.Errorf("Service name is required to initialize tracing")
+		serviceName = "unknown"
+	}
+
+	if config.GetTracingEndpoint() == "" {
+		log.Errorf("Tracing endpoint is required to initialize tracing. Disabling tracing. Please set the endpoint in the config file.")
+		config.SetTracingEnabled(false)
+		return func() {}
+	}
+
+	ctx := context.Background()
+
+	client := otlptracehttp.NewClient(
+		otlptracehttp.WithEndpointURL(config.GetTracingEndpoint()), // Your collector endpoint
+		otlptracehttp.WithInsecure(),                               // Use WithInsecure for non-production environments
+	)
+
+	exporter, err := otlptrace.New(ctx, client)
+	if err != nil {
+		log.Fatalf("Failed to create the trace exporter: %v", err)
+	}
+
+	customExporter := &customExporter{
+		exporter: exporter,
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(serviceName),
+		),
+	)
+	if err != nil {
+		log.Fatalf("Failed to create resource: %v", err)
+	}
+
+	tp := trace.NewTracerProvider(
+		// trace.WithSampler(trace.AlwaysSample()), // Change this to trace.NeverSample() for production
+		trace.WithBatcher(customExporter),
+		trace.WithResource(res),
+	)
+
+	otel.SetTracerProvider(tp)
+
+	return func() {
+		if err := tp.Shutdown(ctx); err != nil {
+			log.Fatalf("Failed to shut down tracer provider: %v", err)
+		}
+	}
```

Comment:
all of the above `Fatalf` maybe too harsh. If for whatever reason tracing export is not able to initialize, we should log a error, disable tracing but still continue bringing up the app. It is ok to not run with tracing rather than not able to start siglens due to some module issue with tracing